hibernate-core/src/main/java/org/hibernate/hql/spi/id/cte/ValuesListCteBasedDeleteHandlerImpl.java

```diff
@@ -0,0 +1,215 @@
+
+	private final String idSelect;
+	private final List<ParameterSpecification> idSelectParameterSpecifications;
+	private final List<String> deletes;
+
+	public ValuesListCteBasedDeleteHandlerImpl(
+			SessionFactoryImplementor factory,
+			HqlSqlWalker walker) {
+		this( factory, walker, null, null );
+	}
+
+	public ValuesListCteBasedDeleteHandlerImpl(
+			SessionFactoryImplementor factory,
+			HqlSqlWalker walker,
+			String catalog,
+			String schema) {
+		super( factory, walker, catalog, schema );
+
+		DeleteStatement deleteStatement = (DeleteStatement) walker.getAST();
+		FromElement fromElement = deleteStatement.getFromClause().getFromElement();
+
+		this.targetedPersister = fromElement.getQueryable();
+
+		final ProcessedWhereClause processedWhereClause = processWhereClause( deleteStatement.getWhereClause() );
+		this.idSelectParameterSpecifications = processedWhereClause.getIdSelectParameterSpecifications();
+
+		final String bulkTargetAlias = fromElement.getTableAlias();
+
+		this.idSelect = generateIdSelect( bulkTargetAlias, processedWhereClause ).toStatementString();
+
+		final String idSubselect = generateIdSubselect( targetedPersister );
+		deletes = new ArrayList<>();
+
+		// If many-to-many, delete the FK row in the collection table.
+		// This partially overlaps with DeleteExecutor, but it instead uses the
+		// temp table in the idSubselect.
+		for ( Type type : targetedPersister.getPropertyTypes() ) {
+			if ( type.isCollectionType() ) {
+				CollectionType cType = (CollectionType) type;
+				AbstractCollectionPersister cPersister = (AbstractCollectionPersister) factory.getCollectionPersister( cType.getRole() );
+				if ( cPersister.isManyToMany() ) {
+					deletes.add( generateDelete(
+							cPersister.getTableName(),
+							cPersister.getKeyColumnNames(),
+							idSubselect,
+							"bulk delete - m2m join table cleanup"
+					) );
+				}
+			}
+		}
+
+		String[] tableNames = targetedPersister
+				.getConstraintOrderedTableNameClosure();
+		String[][] columnNames = targetedPersister
+				.getContraintOrderedTableKeyColumnClosure();
+		for ( int i = 0; i < tableNames.length; i++ ) {
+			// TODO : an optimization here would be to consider cascade deletes and not gen those delete statements;
+			//      the difficulty is the ordering of the tables here vs the cascade attributes on the persisters ->
+			//          the table info gotten here should really be self-contained (i.e., a class representation
+			//          defining all the needed attributes), then we could then get an array of those
+			deletes.add( generateDelete( tableNames[i], columnNames[i], idSubselect, "bulk delete" ) );
+		}
+	}
+
+	private String generateDelete(
+			String tableName,
+			String[] columnNames,
+			String idSubselect,
+			String comment) {
+		final Delete delete = new Delete().setTableName( tableName ).setWhere(
+				"(" + StringHelper.join( ", ", columnNames ) + ") in ("
+						+ idSubselect + ")" );
```

Comment:
I just checked `org.hibernate.hql.spi.id.TableBasedDeleteHandlerImpl.generateDelete(String, String[], String, String)` in latest release of Hibernate and it also uses same multi-column IN clause. If no one complained about it yet, either people don't use hierarchies with multi-column IDs or it works in most databases.